cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java

```diff
@@ -521,23 +522,30 @@ public void removeStreamTags(HttpRequest request, HttpResponder responder,
   }
 
   private Map<String, String> readMetadata(HttpRequest request) throws BadRequestException, IOException {
-    ChannelBuffer content = request.getContent();
-    if (!content.readable()) {
-      throw new BadRequestException("Unable to read business metadata from request.");
-    }
-    try (Reader reader = new InputStreamReader(new ChannelBufferInputStream(content), Charsets.UTF_8)) {
-      return GSON.fromJson(reader, MAP_STRING_STRING_TYPE);
+    try {
+      ChannelBuffer content = request.getContent();
+      Preconditions.checkArgument(content.readable());
+      try (Reader reader = new InputStreamReader(new ChannelBufferInputStream(content), Charsets.UTF_8)) {
+        Map<String, String> metadata = GSON.fromJson(reader, MAP_STRING_STRING_TYPE);
+        Preconditions.checkNotNull(metadata);
+        return metadata;
+      }
+    } catch (Throwable t) {
+      throw new BadRequestException("Unable to read business metadata from the request.", t);
     }
   }
 
   private String[] readArray(HttpRequest request) throws BadRequestException, IOException {
-    ChannelBuffer content = request.getContent();
-    if (!content.readable()) {
-      throw new BadRequestException("Unable to read a list of keys from the request.");
-    }
-    try (Reader reader = new InputStreamReader(new ChannelBufferInputStream(content), Charsets.UTF_8)) {
-      List<String> toReturn = GSON.fromJson(reader, LIST_STRING_TYPE);
-      return toReturn.toArray(new String[toReturn.size()]);
+    try {
+      ChannelBuffer content = request.getContent();
+      Preconditions.checkArgument(content.readable());
```

Comment:
I changed it from that because there are 3 assertions in this method that we need to do and throw BadRequestException if any of those 3 assertions fail:
1) content.isReadable()
2) GSON.fromJson - potential failures, if client doesn't use proper json format
3) GSON.fromJson - returning null

So, instead of individually wrapping each of these things with something like:
```
try {
  GSON.fromJson( ...
catch (JSONException... ) {
  throw new BadRequestException
}
```
I thought its better to wrap the whole method in a try-catch, and throw BadRequestException if any Exception happens while parsing the request body.